gnome-initial-setup/gis-music-player.c

```diff
@@ -0,0 +1,216 @@
+      trans == GST_STATE_CHANGE_READY_TO_PAUSED)
+    {
+      gst_element_seek_simple (priv->playbin, GST_FORMAT_TIME,
+                               (GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_SEGMENT),
+                               0);
+      gst_element_set_state (priv->playbin, GST_STATE_PLAYING);
+    }
+}
+
+static void
+process_tag (const GstTagList *tag_list,
+             const char       *tag,
+             gpointer          user_data)
+{
+  GisMusicPlayer *player = user_data;
+  GisMusicPlayerPrivate *priv = gis_music_player_get_instance_private (player);
+
+  /* loop points are non-standard, so they go in the extended
+   * comments, which is gstreamer's "catch-all" */
+  if (strcmp (tag, GST_TAG_EXTENDED_COMMENT) == 0)
+    {
+      int n = gst_tag_list_get_tag_size (tag_list, tag);
+      while (n--)
+        {
+          char *str;
+          int64_t loop_point;
+
+          if (!gst_tag_list_get_string_index (tag_list, tag, n, &str))
+            continue;
+          if (sscanf (str, "LOOP_POINT=%" G_GINT64_FORMAT, &loop_point) != 0)
+            priv->loop_point = loop_point;
+
+          g_free (str);
+        }
+    }
+}
+
+static void
+on_tag (GstBus     *bus,
+        GstMessage *message,
+        gpointer    user_data)
+{
+  GstTagList *tag_list;
+  gst_message_parse_tag (message, &tag_list);
+  gst_tag_list_foreach (tag_list, process_tag, user_data);
+  gst_tag_list_unref (tag_list);
+}
+
+static void
+on_segment_done (GstBus     *bus,
+                 GstMessage *message,
+                 gpointer    user_data)
+{
+  GisMusicPlayer *player = user_data;
+  GisMusicPlayerPrivate *priv = gis_music_player_get_instance_private (player);
+
+  if (priv->playing &&
+      message->src == GST_OBJECT (priv->playbin))
+    {
+      gst_element_seek_simple (priv->playbin, GST_FORMAT_TIME,
+                               GST_SEEK_FLAG_SEGMENT,
+                               priv->loop_point);
+    }
+}
+
+GisMusicPlayer *
+gis_music_player_new (const char *uri)
+{
+  GisMusicPlayer *player = g_object_new (GIS_TYPE_MUSIC_PLAYER, NULL);
+  GisMusicPlayerPrivate *priv = gis_music_player_get_instance_private (player);
+
+  priv->playbin = gst_element_factory_make ("playbin", "player");
+  g_object_set (priv->playbin, "uri", uri, NULL);
+
+  priv->bus = gst_element_get_bus (priv->playbin);
+  gst_bus_add_signal_watch (priv->bus);
+  g_signal_connect (priv->bus, "message::state-changed",
+                    G_CALLBACK (on_state_changed), player);
+  g_signal_connect (priv->bus, "message::segment-done",
+                    G_CALLBACK (on_segment_done), player);
+  g_signal_connect (priv->bus, "message::tag",
+                    G_CALLBACK (on_tag), player);
+
+  return player;
+}
+
+void
+gis_music_player_play (GisMusicPlayer *player)
+{
+  GisMusicPlayerPrivate *priv = gis_music_player_get_instance_private (player);
+  priv->playing = TRUE;
+  gst_element_set_state (priv->playbin, GST_STATE_PAUSED);
```

Comment:
No, it should be `PAUSED`. See the `on_state_changed` which sets it to PLAYING after the preroll is done. We need to hold off in order to do a proper segment seek.